cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -0,0 +1,217 @@
+    super(RetryStrategies.fromConfiguration(cConf, "tethering.agent."));
+    this.connectionInterval = TimeUnit.SECONDS.toMillis(cConf.getLong(Constants.Tethering.CONNECTION_INTERVAL));
+    this.cConf = cConf;
+    this.transactionRunner = transactionRunner;
+    this.store = store;
+    this.instanceName = cConf.get(Constants.INSTANCE_NAME);
+    this.lastMessageIds = new HashMap<>();
+  }
+
+  @Override
+  protected void doStartUp() throws InstantiationException, IllegalAccessException {
+    Class<? extends RemoteAuthenticator> authClass = cConf.getClass(Constants.Tethering.CLIENT_AUTHENTICATOR_CLASS,
+                                                                    null,
+                                                                    RemoteAuthenticator.class);
+    if (authClass != null) {
+      RemoteAuthenticator.setDefaultAuthenticator(authClass.newInstance());
+    }
+    initializeMessageIds();
+  }
+
+  @Override
+  protected long runTask() {
+    List<PeerInfo> peers;
+    try {
+      peers = store.getPeers().stream()
+        // Ignore peers in REJECTED state.
+        .filter(p -> p.getTetheringStatus() != TetheringStatus.REJECTED)
+        .collect(Collectors.toList());
+    } catch (IOException e) {
+      LOG.warn("Failed to get peer information", e);
+      return connectionInterval;
+    }
+    for (PeerInfo peer : peers) {
+      try {
+        Preconditions.checkArgument(peer.getEndpoint() != null,
+                                    "Peer %s doesn't have an endpoint", peer.getName());
+        String uri = CONNECT_CONTROL_CHANNEL + instanceName;
+        String lastMessageId = lastMessageIds.get(peer);
+        if (lastMessageId != null) {
+          uri = uri + "?messageId=" + URLEncoder.encode(lastMessageId, "UTF-8");
+        }
+        HttpResponse resp = TetheringUtils.sendHttpRequest(HttpMethod.GET, new URI(peer.getEndpoint())
+          .resolve(uri));
+        switch (resp.getResponseCode()) {
+          case HttpURLConnection.HTTP_OK:
+            handleResponse(resp, peer);
+            break;
+          case HttpURLConnection.HTTP_NOT_FOUND:
+            handleNotFound(peer);
+            break;
+          case HttpURLConnection.HTTP_FORBIDDEN:
+            handleForbidden(peer);
+            break;
+          default:
+            LOG.error("Peer {} returned unexpected error code {} body {}",
+                      peer.getName(), resp.getResponseCode(),
+                      resp.getResponseBodyAsString(StandardCharsets.UTF_8));
+        }
+      } catch (Exception e) {
+        LOG.debug("Failed to create control channel to {}", peer, e);
+      }
+    }
+
+    // Update last message ids in the store for all peers
+    TransactionRunners.run(transactionRunner, context -> {
+      for (Map.Entry<String, String> entry : lastMessageIds.entrySet()) {
+        AppMetadataStore.create(context).persistSubscriberState(entry.getKey(), SUBSCRIBER, entry.getValue());
```

Comment:
Move the creation of the AppMetadataStore to outside of the `for` loop

Reply:
done